File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java
package com.datastax.spark.connector.japi;

import com.datastax.spark.connector.ColumnSelector;
import com.datastax.spark.connector.RDDFunctions;
import com.datastax.spark.connector.cql.CassandraConnector;
import com.datastax.spark.connector.writer.RowWriterFactory;
import com.datastax.spark.connector.writer.WriteConf;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.rdd.RDD;
import scala.Tuple2;

/**
 * A Java API wrapper over {@link org.apache.spark.rdd.RDD} to provide Spark Cassandra Connector functionality.
 *
 * <p>To obtain an instance of this wrapper, use one of the factory methods in {@link
 * com.datastax.spark.connector.japi.CassandraJavaUtil} class.</p>
 */
@SuppressWarnings("UnusedDeclaration")
public class RDDJavaFunctions<T> extends RDDAndDStreamCommonJavaFunctions<T> {
    public final RDD<T> rdd;
    private final RDDFunctions<T> rddf;

    RDDJavaFunctions(RDD<T> rdd) {
        this.rdd = rdd;
        this.rddf = new RDDFunctions<>(rdd);
    }

    @Override
    public CassandraConnector defaultConnector() {
        return rddf.connector();
    }

    @Override
    protected SparkConf getConf() {
        return rdd.conf();
    }

    @Override
    protected void saveToCassandra(String keyspace, String table, RowWriterFactory<T> rowWriterFactory,
                                   ColumnSelector columnNames, WriteConf conf, CassandraConnector connector) {
        rddf.saveToCassandra(keyspace, table, columnNames, conf, connector, rowWriterFactory);
    }

    // TODO: Add spanBy. To do this we need to first refactor RDDJavaFunctions (probably translate to Scala),
    // because currently ClassTag information is missing and we can't call RDD#map here nor construct a
    // Scala lambda to pass to Scala API.
}

File: spark-cassandra-connector-java/src/main/scala/com/datastax/spark/connector/japi/RDDJavaFunctions.scala
package com.datastax.spark.connector.japi

import java.lang.{Iterable => JIterable}

import org.apache.spark.api.java.JavaPairRDD

import scala.collection.JavaConversions._
import scala.language.implicitConversions
import scala.reflect.ClassTag

import org.apache.spark.rdd.RDD
import org.apache.spark.SparkContext._
import org.apache.spark.api.java.function.{Function => JFunction}

import com.datastax.spark.connector._
import com.datastax.spark.connector.cql.CassandraConnector
import com.datastax.spark.connector.writer.RowWriterFactory
import com.datastax.spark.connector.writer.WriteConf



/**
 * A Java API wrapper over [[org.apache.spark.rdd.RDD]] to provide Spark Cassandra Connector functionality.
 *
 * To obtain an instance of this wrapper, use one of the factory methods in
 * [[com.datastax.spark.connector.japi.CassandraJavaUtil]] class.<
 */

class RDDJavaFunctions[T](val rdd: RDD[T]) extends RDDAndDStreamCommonJavaFunctions[T] {

  override def defaultConnector() = rdd.connector
  override def getConf = rdd.sparkContext.getConf

  override def saveToCassandra(keyspace: String,
                               table: String,
                               rowWriterFactory: RowWriterFactory[T],
                               columnNames: ColumnSelector,
                               conf: WriteConf,
                               connector: CassandraConnector) = {

    rdd.saveToCassandra(keyspace, table, columnNames, conf)(connector, rowWriterFactory)
  }

  override def saveToCassandra(keyspace: String,
                               table: String,
                               cluster: String,
                               rowWriterFactory: RowWriterFactory[T],
                               columnNames: ColumnSelector,
                               conf: WriteConf,
                               connector: CassandraConnector) = {

    rdd.saveToCassandra(keyspace, table, columnNames, conf, Option(cluster))(connector, rowWriterFactory)
  }

  /** Applies a function to each item, and groups consecutive items having the same value together.
    * Contrary to `groupBy`, items from the same group must be already next to each other in the
    * original collection. Works locally on each partition, so items from different
    * partitions will never be placed in the same group.*/
  def spanBy[U : ClassTag](f: JFunction[T, U]): JavaPairRDD[U, JIterable[T]] = {
    new JavaPairRDD(rdd.spanBy(f.call).mapValues(asJavaIterable))
  }
}

